src/json-patch-duplex.ts

```diff
@@ -524,6 +525,74 @@ module jsonpatch {
     return true;
   }
 
+  
+  /**
+     * Apply a single json-patch on an object tree
+     * Returns the result object.
+     */
+  export function applyPatch<T>(tree: T, patch: Patch<any>): T;
+  export function applyPatch<T>(tree: T, patch: Patch<any>, validate: boolean): T;
+  export function applyPatch<T>(tree: T, patch: Patch<any>, validate = false): T {
+    if (typeof validate !== 'boolean') {
+      validate = false;
+    }
+    // on root
+    if (patch.path === "") {
+      if (patch.op === 'add' || patch.op === 'replace') {// same for add or replace 
+        // a primitive value, just return it
+        if (typeof patch.value !== 'object' || patch.value === null) {
+          return patch.value;
+        } else {
+          // conflicting types [] vs {} just return the new value
+          if (_isArray(tree) !== _isArray(patch.value)) {
+            return patch.value;
+            //same type, use original apply
```

Comment:
This is not a big issue, but why are we making the effort here to mutate the original object at all if the function returns a new reference in some cases anyway? The new object is already in the heap. It is way faster to just switch the reference out than mutating the original object, basically diffing it, removing all properties, then copying over all from the new object. Given the performance focus of this package I would vote for always returning the new value if the root is replaced (it also makes more sense semantically).